// @visactor/vseed/dist/esm/dataReshape/constant.js
const ORIGINAL_DATA = '__OriginalData__';
const Separator = '-';
const FoldMeasureName = '__MeaName__';
const FoldMeasureId = '__MeaId__';
const FoldMeasureValue = '__MeaValue__';
const MeasureId = FoldMeasureId;
const MeasureName = FoldMeasureName;
const FoldPrimaryMeasureValue = '__MeaPrimaryValue__';
const FoldSecondaryMeasureValue = '__MeaSecondaryValue__';
const FoldXMeasureValue = '__MeaXValue__';
const FoldYMeasureValue = '__MeaYValue__';
const FoldXMeasureId = '__MeaXId__';
const FoldYMeasureId = '__MeaYId__';
const XEncoding = '__Dim_X__';
const YEncoding = '__Dim_Y__';
const AngleEncoding = '__Dim_Angle__';
const DetailEncoding = '__Dim_Detail__';
const ColorEncoding = '__Dim_Color__';
const ColorIdEncoding = '__Dim_ColorId__';
const Q1MeasureValue = '__Q1__';
const Q3MeasureValue = '__Q3__';
const LowerWhisker = '__Lower_Whisker__';
const UpperWhisker = '__Upper_Whisker__';
const OutliersMeasureId = '__Outliers__';
const MedianMeasureId = '__Meadian__';
const BinStartMeasureId = '__BinStart__';
const BinEndMeasureId = '__BinEnd__';
const BinCountMeasureId = '__BinCount__';
const BinPercentageMeasureId = '__BinPercentage__';
export { AngleEncoding, BinCountMeasureId, BinEndMeasureId, BinPercentageMeasureId, BinStartMeasureId, ColorEncoding, ColorIdEncoding, DetailEncoding, FoldMeasureId, FoldMeasureName, FoldMeasureValue, FoldPrimaryMeasureValue, FoldSecondaryMeasureValue, FoldXMeasureId, FoldXMeasureValue, FoldYMeasureId, FoldYMeasureValue, LowerWhisker, MeasureId, MeasureName, MedianMeasureId, ORIGINAL_DATA, OutliersMeasureId, Q1MeasureValue, Q3MeasureValue, Separator, UpperWhisker, XEncoding, YEncoding };
