// @visactor/vseed/dist/esm/dataReshape/dataReshapeByEncoding.js
import { foldMeasures } from "./foldMeasures.js";
import { FoldMeasureId, FoldMeasureName, FoldMeasureValue, Separator } from "./constant.js";
import { unfoldDimensions } from "./unfoldDimensions.js";
const dataReshapeByEncoding = (dataset, dimensions, measures, encoding, options)=>{
    const { foldMeasureId = FoldMeasureId, foldMeasureName = FoldMeasureName, foldMeasureValue = FoldMeasureValue, colorItemAsId = false, colorMeasureId, omitIds } = options || {};
    const { dataset: foldedDataset, foldInfo } = foldMeasures(dataset, measures, {
        measureId: foldMeasureId,
        measureName: foldMeasureName,
        measureValue: foldMeasureValue,
        colorMeasureId,
        omitIds
    });
    const { dataset: finalDataset, unfoldInfo } = unfoldDimensions(foldedDataset, dimensions, encoding, {
        foldMeasureId,
        separator: Separator,
        colorItemAsId
    });
    return {
        dataset: finalDataset,
        foldInfo,
        unfoldInfo
    };
};
export { dataReshapeByEncoding };
