// @visactor/vseed/dist/esm/pipeline/spec/chart/pipes/markStyle/areaStyle.js
import { selector } from "../../../../../dataSelector/index.js";
import { groupBy, isEmpty, isNullish } from "remeda";
import { getCurveTension, getCurveType } from "./curve/index.js";
const areaStyle_areaStyle = (spec, context)=>{
    const { advancedVSeed } = context;
    const { markStyle, datasetReshapeInfo, dataset } = advancedVSeed;
    const { areaStyle, lineStyle } = markStyle;
    const { unfoldInfo } = datasetReshapeInfo[0];
    const result = {
        ...spec,
        area: {
            visible: true,
            style: {}
        }
    };
    if (isNullish(areaStyle) || isEmpty(areaStyle)) return result;
    const areaStyles = Array.isArray(areaStyle) ? areaStyle : [
        areaStyle
    ];
    const lineStyles = Array.isArray(lineStyle) ? lineStyle : [
        lineStyle
    ];
    const group = unfoldInfo.encodingColorId;
    const areaGroups = groupBy(dataset, (d)=>d[group ?? '']);
    const customMap = areaStyles.reduce((result, style, index)=>{
        const { areaColor, areaColorOpacity, areaVisible = true } = style;
        const curveType = getCurveType(context.vseed, lineStyles[index]?.lineSmooth);
        const curveTension = getCurveTension(context.vseed, lineStyles[index]?.lineSmooth);
        return {
            ...result,
            [`custom${index + 1}`]: {
                level: index + 1,
                filter: (datum)=>{
                    const lineData = areaGroups[datum[group ?? '']];
                    for (const d of lineData)if (selector(d, style.selector)) return true;
                    return false;
                },
                style: {
                    curveType,
                    curveTension,
                    visible: areaVisible,
                    fill: areaColor,
                    fillOpacity: areaColorOpacity
                }
            }
        };
    }, {});
    return {
        ...result,
        area: {
            ...result.area,
            visible: true,
            state: {
                ...customMap
            }
        }
    };
};
export { areaStyle_areaStyle as areaStyle };
