// @visactor/vseed/dist/esm/pipeline/spec/chart/pipes/markStyle/outlierStyle.js
import { selector } from "../../../../../dataSelector/index.js";
import { isEmpty, isNullish } from "remeda";
const outlierStyle_outlierStyle = (spec, context)=>{
    const { advancedVSeed, vseed } = context;
    const { markStyle, config } = advancedVSeed;
    const { outlierStyle } = markStyle;
    const theme = config?.[vseed.chartType]?.outlierStyle;
    const result = {
        ...spec,
        outlier: {
            style: {
                fill: theme?.pointColor
            }
        }
    };
    if (isNullish(outlierStyle) || isEmpty(outlierStyle)) return result;
    const outlierStyles = Array.isArray(outlierStyle) ? outlierStyle : [
        outlierStyle
    ];
    const customMap = outlierStyles.reduce((result, style, index)=>{
        const { pointBorderColor, pointBorderStyle, pointBorderWidth = 1, pointColor, pointColorOpacity, pointSize, pointVisible = true } = style;
        const lineDash = 'dashed' === pointBorderStyle ? [
            5,
            2
        ] : 'dotted' === pointBorderStyle ? [
            2,
            5
        ] : [
            0,
            0
        ];
        return {
            ...result,
            [`custom${index + 1}`]: {
                level: index + 1,
                filter: (datum)=>{
                    if (selector(datum, style.selector)) return true;
                    return false;
                },
                style: {
                    visible: pointVisible,
                    size: pointSize,
                    fill: pointColor,
                    fillOpacity: pointColorOpacity,
                    innerBorder: {
                        stroke: pointBorderColor,
                        lineWidth: pointBorderWidth,
                        distance: (pointBorderWidth || 0) / 2,
                        lineDash: lineDash
                    }
                }
            }
        };
    }, {});
    return {
        ...result,
        outlier: {
            ...result.outlier,
            state: {
                ...customMap
            }
        }
    };
};
export { outlierStyle_outlierStyle as outlierStyle };
