// @visactor/vseed/dist/esm/pipeline/spec/chart/pipes/markStyle/pointState.js
const pointStateDimensionHover = (spec)=>{
    const point = spec.point || {};
    const result = {
        ...spec,
        point: {
            ...point,
            state: {
                ...point.state || {},
                dimension_hover: {
                    scaleX: 1.4,
                    scaleY: 1.4,
                    outerBorder: {
                        lineWidth: 4,
                        strokeOpacity: 0.25,
                        distance: 2
                    }
                }
            }
        }
    };
    return result;
};
const pointStateHover = (spec, context)=>{
    const point = spec.point || {};
    const { advancedVSeed } = context;
    const { datasetReshapeInfo } = advancedVSeed;
    const { unfoldInfo } = datasetReshapeInfo[0];
    const result = {
        ...spec,
        point: {
            ...point,
            state: {
                ...point.state || {},
                hover: {
                    scaleX: 1.4,
                    scaleY: 1.4,
                    stroke: (datum, context)=>{
                        const field = unfoldInfo.encodingColorId;
                        const color = context.seriesColor(datum[field]);
                        return color;
                    },
                    fillOpacity: 0.6,
                    lineWidth: 1
                }
            }
        }
    };
    return result;
};
export { pointStateDimensionHover, pointStateHover };
