// @visactor/vseed/dist/esm/pipeline/advanced/chart/pipes/reshape/pivotReshapeWithScatterEncoding.js
import { FoldXMeasureValue, FoldYMeasureValue, dataReshapeByEncoding } from "../../../../../dataReshape/index.js";
import { getColorMeasureId } from "../../../../spec/chart/pipes/index.js";
import { findAllMeasures, measureDepth } from "../../../../utils/index.js";
const pivotReshapeWithScatterEncoding = (advancedVSeed, context)=>{
    const result = {
        ...advancedVSeed
    };
    const { vseed } = context;
    const { dataset } = vseed;
    const { encoding, chartType } = advancedVSeed;
    const measures = advancedVSeed.reshapeMeasures ?? advancedVSeed.measures ?? [];
    const dimensions = advancedVSeed.reshapeDimensions ?? advancedVSeed.dimensions ?? [];
    const allMeasures = findAllMeasures(measures);
    const measureGroups = [];
    const depth = measureDepth(measures);
    if (3 === depth) measures.forEach((measure)=>{
        measureGroups.push(measure.children);
    });
    else if (2 === depth) measureGroups.push(measures);
    const datasetList = [];
    const datasetReshapeInfo = [];
    measureGroups.forEach((measures, index)=>{
        if (measures.length > 2) throw new Error('measures can not be more than 2 groups in scatter');
        const foldInfoList = [];
        const unfoldInfoList = [];
        const datasets = [];
        const xMeasures = measures[0];
        const yMeasures = measures[1] || xMeasures;
        if (xMeasures && xMeasures.children) {
            const { dataset: newDataset, foldInfo, unfoldInfo } = dataReshapeByEncoding(dataset, dimensions, xMeasures.children, encoding, {
                foldMeasureValue: `${FoldXMeasureValue}${index}`,
                colorItemAsId: true,
                colorMeasureId: getColorMeasureId(advancedVSeed, vseed),
                omitIds: allMeasures.map((item)=>item.id)
            });
            datasets.push(newDataset);
            foldInfoList.push(foldInfo);
            unfoldInfoList.push(unfoldInfo);
        }
        if (yMeasures && yMeasures.children) {
            const { dataset: newDataset, foldInfo, unfoldInfo } = dataReshapeByEncoding(dataset, dimensions, yMeasures.children, encoding, {
                foldMeasureValue: `${FoldYMeasureValue}${index}`,
                colorItemAsId: true,
                colorMeasureId: getColorMeasureId(advancedVSeed, vseed),
                omitIds: allMeasures.map((item)=>item.id)
            });
            datasets.push(newDataset);
            foldInfoList.push(foldInfo);
            unfoldInfoList.push(unfoldInfo);
        }
        const unfoldInfo = {
            ...unfoldInfoList[0],
            colorItems: unfoldInfoList.flatMap((d)=>d.colorItems),
            colorIdMap: unfoldInfoList.reduce((prev, cur)=>({
                    ...prev,
                    ...cur.colorIdMap
                }), {})
        };
        const reshapeInfo = {
            id: `${chartType}-${index}`,
            index,
            foldInfo: foldInfoList[0],
            foldInfoList: foldInfoList,
            unfoldInfo: unfoldInfo
        };
        datasetReshapeInfo.push(reshapeInfo);
        datasetList.push(datasets[0].map((d, index)=>({
                ...d,
                ...datasets[1]?.[index] || {}
            })));
    });
    return {
        ...result,
        dataset: datasetList,
        datasetReshapeInfo: datasetReshapeInfo
    };
};
export { pivotReshapeWithScatterEncoding };
