// @visactor/vseed/dist/esm/pipeline/advanced/chart/pipes/reshape/pivotReshapeWithDualEncoding.js
import { uniqueBy } from "remeda";
import { FoldPrimaryMeasureValue, FoldSecondaryMeasureValue, dataReshapeByEncoding } from "../../../../../dataReshape/index.js";
import { getColorMeasureId } from "../../../../spec/chart/pipes/index.js";
import { findAllMeasures, measureDepth } from "../../../../utils/index.js";
const pivotReshapeWithDualEncoding = (advancedVSeed, context)=>{
    const result = {
        ...advancedVSeed
    };
    const { vseed } = context;
    const { dataset } = vseed;
    const { encoding, chartType } = advancedVSeed;
    const measures = advancedVSeed.reshapeMeasures ?? advancedVSeed.measures ?? [];
    const dimensions = advancedVSeed.reshapeDimensions ?? advancedVSeed.dimensions ?? [];
    const allMeasures = findAllMeasures(measures);
    const datasetList = [];
    const datasetReshapeInfo = [];
    const measureGroups = [];
    const depth = measureDepth(measures);
    if (3 === depth) measures.forEach((measure)=>{
        measureGroups.push(measure.children);
    });
    else if (2 === depth) measureGroups.push(measures);
    measureGroups.forEach((measures, index)=>{
        if (measures.length > 2) throw new Error('measures can not be more than 2 groups in dualAxis');
        const foldInfoList = [];
        const unfoldInfoList = [];
        const datasets = [];
        const primaryMeasures = measures[0];
        const secondaryMeasures = measures[1] || [];
        if (primaryMeasures && primaryMeasures.children) {
            const { dataset: newDataset, foldInfo, unfoldInfo } = dataReshapeByEncoding(dataset, uniqueBy(dimensions, (item)=>item.id), uniqueBy(primaryMeasures.children, (item)=>item.id), encoding, {
                colorItemAsId: false,
                foldMeasureValue: `${FoldPrimaryMeasureValue}${index}`,
                colorMeasureId: getColorMeasureId(advancedVSeed, vseed),
                omitIds: allMeasures.map((item)=>item.id)
            });
            datasets.push(newDataset);
            foldInfoList.push(foldInfo);
            unfoldInfoList.push(unfoldInfo);
        }
        if (secondaryMeasures && secondaryMeasures.children) {
            const { dataset: newDataset, foldInfo, unfoldInfo } = dataReshapeByEncoding(dataset, uniqueBy(dimensions, (item)=>item.id), uniqueBy(secondaryMeasures.children, (item)=>item.id), encoding, {
                colorItemAsId: false,
                foldMeasureValue: `${FoldSecondaryMeasureValue}${index}`,
                colorMeasureId: getColorMeasureId(advancedVSeed, vseed),
                omitIds: allMeasures.map((item)=>item.id)
            });
            datasets.push(newDataset);
            foldInfoList.push(foldInfo);
            unfoldInfoList.push(unfoldInfo);
        }
        const unfoldInfo = {
            ...unfoldInfoList[0],
            colorItems: unfoldInfoList.flatMap((d)=>d.colorItems),
            colorIdMap: unfoldInfoList.reduce((prev, cur)=>({
                    ...prev,
                    ...cur.colorIdMap
                }), {})
        };
        const reshapeInfo = {
            id: `${chartType}-${index}`,
            index,
            foldInfo: foldInfoList[0],
            foldInfoList: foldInfoList,
            unfoldInfo: unfoldInfo
        };
        datasetReshapeInfo.push(reshapeInfo);
        datasetList.push(datasets.flat(2));
    });
    return {
        ...result,
        dataset: datasetList,
        datasetReshapeInfo: datasetReshapeInfo
    };
};
export { pivotReshapeWithDualEncoding };
