// @visactor/vseed/dist/esm/pipeline/advanced/chart/pipes/reshape/reshapeWithScatterEncoding.js
import { uniqueBy } from "remeda";
import { FoldXMeasureId, FoldXMeasureValue, FoldYMeasureId, FoldYMeasureValue, dataReshapeByEncoding } from "../../../../../dataReshape/index.js";
import { getColorMeasureId } from "../../../../spec/chart/pipes/index.js";
const reshapeWithScatterEncoding = (advancedVSeed, context)=>{
    const result = {
        ...advancedVSeed
    };
    const { vseed } = context;
    const { dataset } = vseed;
    const { encoding, chartType } = advancedVSeed;
    const measures = advancedVSeed.reshapeMeasures ?? advancedVSeed.measures ?? [];
    const dimensions = advancedVSeed.reshapeDimensions ?? advancedVSeed.dimensions ?? [];
    if (measures.length > 2) throw new Error('measures can not be more than 2 groups in scatter');
    const foldInfoList = [];
    const unfoldInfoList = [];
    const datasets = [];
    const xMeasures = measures[0];
    const yMeasures = measures[1] || xMeasures;
    if (xMeasures && xMeasures.children) {
        const { dataset: newDataset, foldInfo, unfoldInfo } = dataReshapeByEncoding(dataset, uniqueBy(dimensions, (d)=>d.id), uniqueBy(xMeasures.children, (d)=>d.id), encoding, {
            foldMeasureValue: FoldXMeasureValue,
            foldMeasureId: FoldXMeasureId,
            colorItemAsId: true,
            colorMeasureId: getColorMeasureId(advancedVSeed, vseed)
        });
        datasets.push(newDataset);
        foldInfoList.push(foldInfo);
        unfoldInfoList.push(unfoldInfo);
    }
    if (yMeasures && yMeasures.children) {
        const { dataset: newDataset, foldInfo, unfoldInfo } = dataReshapeByEncoding(datasets[0], uniqueBy(dimensions, (d)=>d.id), uniqueBy(yMeasures.children, (d)=>d.id), encoding, {
            foldMeasureValue: FoldYMeasureValue,
            foldMeasureId: FoldYMeasureId,
            colorItemAsId: true,
            colorMeasureId: getColorMeasureId(advancedVSeed, vseed)
        });
        datasets[0] = newDataset;
        foldInfoList.push(foldInfo);
        unfoldInfoList.push(unfoldInfo);
    }
    const unfoldInfo = {
        ...unfoldInfoList[0],
        colorItems: unfoldInfoList.flatMap((d)=>d.colorItems),
        colorIdMap: unfoldInfoList.reduce((prev, cur)=>({
                ...prev,
                ...cur.colorIdMap
            }), {})
    };
    return {
        ...result,
        dataset: datasets[0],
        datasetReshapeInfo: [
            {
                id: String(chartType),
                index: 0,
                foldInfo: foldInfoList[0],
                foldInfoList: foldInfoList,
                unfoldInfo: unfoldInfo
            }
        ]
    };
};
export { reshapeWithScatterEncoding };
