// @visactor/vseed/dist/esm/pipeline/advanced/chart/pipes/reshape/reshapeWithDualEncoding.js
import { uniqueBy } from "remeda";
import { FoldPrimaryMeasureValue, FoldSecondaryMeasureValue, dataReshapeByEncoding } from "../../../../../dataReshape/index.js";
import { getColorMeasureId } from "../../../../spec/chart/pipes/index.js";
const reshapeWithDualEncoding = (advancedVSeed, context)=>{
    const result = {
        ...advancedVSeed
    };
    const { vseed } = context;
    const { dataset } = vseed;
    const { encoding, chartType } = advancedVSeed;
    const measures = advancedVSeed.reshapeMeasures ?? advancedVSeed.measures ?? [];
    const dimensions = advancedVSeed.reshapeDimensions ?? advancedVSeed.dimensions ?? [];
    if (measures.length > 2) throw new Error('measures can not be more than 2 groups in dualAxis');
    const foldInfoList = [];
    const unfoldInfoList = [];
    const datasets = [];
    const primaryMeasures = measures[0];
    const secondaryMeasures = measures[1] || [];
    if (primaryMeasures && primaryMeasures.children) {
        const { dataset: newDataset, foldInfo, unfoldInfo } = dataReshapeByEncoding(dataset, uniqueBy(dimensions, (item)=>item.id), uniqueBy(primaryMeasures.children, (item)=>item.id), encoding, {
            colorItemAsId: false,
            foldMeasureValue: FoldPrimaryMeasureValue,
            colorMeasureId: getColorMeasureId(advancedVSeed, vseed)
        });
        datasets.push(newDataset);
        foldInfoList.push(foldInfo);
        unfoldInfoList.push(unfoldInfo);
    }
    if (secondaryMeasures && secondaryMeasures.children) {
        const { dataset: newDataset, foldInfo, unfoldInfo } = dataReshapeByEncoding(dataset, uniqueBy(dimensions, (item)=>item.id), uniqueBy(secondaryMeasures.children, (item)=>item.id), encoding, {
            colorItemAsId: false,
            foldMeasureValue: FoldSecondaryMeasureValue,
            colorMeasureId: getColorMeasureId(advancedVSeed, vseed)
        });
        datasets.push(newDataset);
        foldInfoList.push(foldInfo);
        unfoldInfoList.push(unfoldInfo);
    }
    const unfoldInfo = {
        ...unfoldInfoList[0],
        colorItems: unfoldInfoList.flatMap((d)=>d.colorItems),
        colorIdMap: unfoldInfoList.reduce((prev, cur)=>({
                ...prev,
                ...cur.colorIdMap
            }), {})
    };
    return {
        ...result,
        dataset: datasets,
        datasetReshapeInfo: [
            {
                id: String(chartType),
                index: 0,
                foldInfo: foldInfoList[0],
                foldInfoList: foldInfoList,
                unfoldInfo: unfoldInfo
            }
        ]
    };
};
export { reshapeWithDualEncoding };
