// @visactor/vseed/dist/esm/pipeline/spec/chart/pipes/axes/xBand.js
import { defaultTitleText } from "./title/defaultTitleText.js";
import { AXIS_LABEL_SPACE } from "../../../../utils/index.js";
import { MeasureId } from "../../../../../dataReshape/index.js";
import { isArray } from "@visactor/vutils";
const xBand = (spec, context)=>{
    const result = {
        ...spec
    };
    const { advancedVSeed, vseed } = context;
    const { chartType } = vseed;
    const { measures, dimensions, encoding, datasetReshapeInfo, pivotAllDatasetReshapeInfo } = advancedVSeed;
    const config = advancedVSeed.config?.[chartType]?.xAxis ?? {};
    if (!result.axes) result.axes = [];
    const { visible = true, label, tick, title, grid, line, labelAutoHide, labelAutoHideGap, labelAutoLimit, labelAutoLimitLength = 80, labelAutoRotate, labelAutoRotateAngleRange } = config;
    const sampling = !(labelAutoHide || labelAutoRotate || labelAutoLimit);
    const onlyMeasureId = 0 === (encoding.x || []).filter((v)=>v !== MeasureId).length;
    const bandAxis = {
        visible,
        type: 'band',
        orient: 'bottom',
        maxHeight: labelAutoLimitLength + 60,
        sampling,
        hover: true,
        label: {
            visible: label?.visible,
            flush: true,
            space: AXIS_LABEL_SPACE,
            style: {
                maxLineWidth: labelAutoLimitLength,
                fill: label?.labelColor,
                angle: label?.labelAngle,
                fontSize: label?.labelFontSize,
                fontWeight: label?.labelFontWeight
            },
            minGap: labelAutoHideGap,
            autoHide: labelAutoHide,
            autoHideMethod: 'greedy',
            autoHideSeparation: labelAutoHideGap,
            autoLimit: labelAutoLimit,
            autoRotate: labelAutoRotate,
            autoRotateAngle: labelAutoRotateAngleRange,
            lastVisible: true
        },
        title: {
            visible: title?.visible,
            text: title?.titleText || defaultTitleText(measures, dimensions, encoding.x),
            style: {
                fill: title?.titleColor,
                fontSize: title?.titleFontSize,
                fontWeight: title?.titleFontWeight
            }
        },
        tick: {
            visible: tick?.visible,
            tickSize: tick?.tickSize,
            inside: tick?.tickInside,
            style: {
                stroke: tick?.tickColor
            }
        },
        grid: {
            visible: grid?.visible,
            style: {
                lineWidth: grid?.gridWidth,
                stroke: grid?.gridColor,
                lineDash: grid?.gridLineDash
            }
        },
        domainLine: {
            visible: line?.visible,
            style: {
                lineWidth: line?.lineWidth,
                stroke: line?.lineColor
            }
        },
        paddingInner: [
            0.15,
            0.1
        ],
        paddingOuter: [
            0.075,
            0.1
        ]
    };
    if (onlyMeasureId && bandAxis.label) {
        const allDatasetReshapeInfo = pivotAllDatasetReshapeInfo || datasetReshapeInfo;
        const colorIdMap = allDatasetReshapeInfo.reduce((prev, cur)=>({
                ...prev,
                ...cur.unfoldInfo.colorIdMap
            }), {});
        bandAxis.label.formatMethod = (text)=>isArray(text) ? text : colorIdMap[String(text)]?.alias ?? text;
    }
    result.axes = [
        ...result.axes,
        bandAxis
    ];
    return result;
};
export { xBand };
