// @visactor/vseed/dist/esm/theme/common/annotaion.js
const getDefaultAnnotationPoint = ()=>({
        textBackgroundVisible: true,
        textFontSize: 12,
        textFontWeight: 400,
        textBackgroundBorderRadius: 0,
        textBackgroundBorderWidth: 1,
        textBackgroundPadding: 2
    });
const getLightAnnotationPoint = ()=>({
        ...getDefaultAnnotationPoint(),
        textColor: '#ffffff',
        textBackgroundColor: '#BCC1CB',
        textBackgroundBorderColor: '#BCC1CB'
    });
const getDefaultAnnotationLine = ()=>({
        lineVisible: true,
        lineWidth: 1,
        lineStyle: 'solid',
        textFontSize: 12,
        textFontWeight: 400,
        startSymbolVisible: true,
        endSymbolVisible: false,
        textBackgroundVisible: true,
        textBackgroundBorderWidth: 1,
        textBackgroundBorderRadius: 0,
        textBackgroundPadding: 2
    });
const getLightAnnotationHorizontalLine = ()=>({
        ...getDefaultAnnotationLine(),
        lineColor: '#BCC1CB',
        textColor: '#ffffff',
        textBackgroundColor: '#BCC1CB',
        textBackgroundBorderColor: '#BCC1CB'
    });
const getLightAnnotationVerticalLine = ()=>getLightAnnotationHorizontalLine();
const getDefaultAnnotationArea = ()=>({
        textFontSize: 12,
        textFontWeight: 400,
        textBackgroundVisible: true,
        textBackgroundBorderWidth: 1,
        textBackgroundBorderRadius: 0,
        textBackgroundPadding: 2,
        areaBorderWidth: 0,
        areaBorderRadius: 0,
        areaLineDash: [
            2,
            2
        ],
        outerPadding: 4
    });
const getLightAnnotationArea = ()=>({
        ...getDefaultAnnotationArea(),
        textColor: '#ffffff',
        textBackgroundColor: '#BCC1CB',
        textBackgroundBorderColor: '#BCC1CB',
        areaColor: '#BCC1CB',
        areaColorOpacity: 0.12,
        areaBorderColor: '#BCC1CB'
    });
const getLightAnnotation = ()=>({
        annotationPoint: getLightAnnotationPoint(),
        annotationHorizontalLine: getLightAnnotationHorizontalLine(),
        annotationVerticalLine: getLightAnnotationVerticalLine(),
        annotationArea: getLightAnnotationArea()
    });
const getDarkAnnotationPoint = ()=>({
        ...getDefaultAnnotationPoint(),
        textColor: '#E2E3E6',
        textBackgroundColor: '#55595F',
        textBackgroundBorderColor: '#55595F'
    });
const getDarkAnnotationHorizontalLine = ()=>({
        ...getDefaultAnnotationLine(),
        lineColor: '#55595F',
        textColor: '#E2E3E6',
        textBackgroundColor: '#55595F',
        textBackgroundBorderColor: '#55595F'
    });
const getDarkAnnotationVerticalLine = ()=>getDarkAnnotationHorizontalLine();
const getDarkAnnotationArea = ()=>({
        ...getDefaultAnnotationArea(),
        textColor: '#E2E3E6',
        textBackgroundColor: '#55595F',
        textBackgroundBorderColor: '#55595F',
        areaColor: '#E2E3E6',
        areaColorOpacity: 0.15,
        areaBorderColor: '#E2E3E6'
    });
const getDarkAnnotation = ()=>({
        annotationPoint: getDarkAnnotationPoint(),
        annotationHorizontalLine: getDarkAnnotationHorizontalLine(),
        annotationVerticalLine: getDarkAnnotationVerticalLine(),
        annotationArea: getDarkAnnotationArea()
    });
const annotaion = getLightAnnotation;
export { annotaion as default, getDarkAnnotation, getDarkAnnotationArea, getDarkAnnotationHorizontalLine, getDarkAnnotationPoint, getDarkAnnotationVerticalLine, getDefaultAnnotationArea, getDefaultAnnotationLine, getDefaultAnnotationPoint, getLightAnnotation, getLightAnnotationArea, getLightAnnotationHorizontalLine, getLightAnnotationPoint, getLightAnnotationVerticalLine };
