// @visactor/vseed/dist/esm/theme/common/regressionLine.js
const getDefaultRegressionLine = ()=>({
        lineWidth: 2,
        textFontSize: 12,
        textFontWeight: 400
    });
const getLightRegressionLine = ()=>({
        kdeRegressionLine: {
            ...getDefaultRegressionLine(),
            textColor: '#364159'
        },
        ecdfRegressionLine: {
            ...getDefaultRegressionLine(),
            textColor: '#364159'
        },
        linearRegressionLine: {
            ...getDefaultRegressionLine(),
            textColor: '#364159',
            confidenceIntervalOpacity: 0.2
        },
        lowessRegressionLine: {
            ...getDefaultRegressionLine(),
            textColor: '#364159',
            confidenceIntervalOpacity: 0.2
        },
        polynomialRegressionLine: {
            ...getDefaultRegressionLine(),
            textColor: '#364159',
            confidenceIntervalOpacity: 0.2
        },
        logisticRegressionLine: {
            ...getDefaultRegressionLine(),
            textColor: '#364159',
            confidenceIntervalOpacity: 0.2
        }
    });
const getDarkRegressionLine = ()=>({
        kdeRegressionLine: {
            ...getDefaultRegressionLine(),
            textColor: '#FFFFFF'
        },
        ecdfRegressionLine: {
            ...getDefaultRegressionLine(),
            textColor: '#FFFFFF'
        },
        linearRegressionLine: {
            ...getDefaultRegressionLine(),
            textColor: '#FFFFFF',
            confidenceIntervalOpacity: 0.2
        },
        lowessRegressionLine: {
            ...getDefaultRegressionLine(),
            textColor: '#FFFFFF',
            confidenceIntervalOpacity: 0.2
        },
        polynomialRegressionLine: {
            ...getDefaultRegressionLine(),
            textColor: '#FFFFFF',
            confidenceIntervalOpacity: 0.2
        },
        logisticRegressionLine: {
            ...getDefaultRegressionLine(),
            textColor: '#FFFFFF',
            confidenceIntervalOpacity: 0.2
        }
    });
export { getDarkRegressionLine, getDefaultRegressionLine, getLightRegressionLine };
