// @visactor/vseed/dist/esm/dataSelector/selector.js
import { omit } from "remeda";
function nearlyEqual(a, b, epsilon = 1e-8) {
    if (Number.isNaN(a) || Number.isNaN(b)) return false;
    if (a === b) return true;
    const diff = Math.abs(a - b);
    return diff <= epsilon;
}
const selector_selector = (vchartDatum, selector, selectorMode = 'And')=>{
    if (!selector) return true;
    const vchartKeys = Object.keys(vchartDatum).filter((k)=>k.toLocaleLowerCase().startsWith('__vchart'));
    const datum = omit(vchartDatum, vchartKeys);
    const selectors = Array.isArray(selector) ? selector : [
        selector
    ];
    return selectors['And' === selectorMode ? 'every' : 'some']((selector)=>{
        if (isValueSelector(selector)) return selectByValue(selector, datum);
        if (isMeasureSelector(selector)) return selectByMeasure(selector, datum);
        if (isDimensionSelector(selector)) return selectByDmension(selector, datum);
        if (isPartialDatumSelector(selector)) return selectByPartial(selector, datum);
        return false;
    });
};
const selectorDatum = (datum, selector)=>{
    if (!selector) return [];
    const selectors = Array.isArray(selector) ? selector : [
        selector
    ];
    let finalResult = [];
    for (const selector of selectors){
        const results = [];
        if (isValueSelector(selector)) Object.entries(datum).forEach(([key, value])=>{
            if (value === selector) results.push({
                [key]: value
            });
        });
        else if (isMeasureSelector(selector) && selectByMeasure(selector, datum)) results.push({
            [selector.field]: datum[selector.field]
        });
        else if (isDimensionSelector(selector) && selectByDmension(selector, datum)) results.push({
            [selector.field]: datum[selector.field]
        });
        else if (isPartialDatumSelector(selector) && selectByPartial(selector, datum)) results.push(selector);
        if (results.length) finalResult = finalResult.length ? finalResult.flatMap((prev)=>results.map((r)=>({
                    ...prev,
                    ...r
                }))) : results;
        else {
            finalResult = [];
            break;
        }
    }
    return finalResult;
};
const isValueSelector = (selector)=>'string' == typeof selector || 'number' == typeof selector;
const isPartialDatumSelector = (selector)=>'object' == typeof selector && null !== selector;
const isMeasureSelector = (selector)=>'object' == typeof selector && null !== selector && 'field' in selector && ('operator' in selector || 'op' in selector) && 'value' in selector && ([
        '=',
        '==',
        '!=',
        '>',
        '<',
        '>=',
        '<=',
        'between'
    ].includes(selector.operator) || [
        '=',
        '==',
        '!=',
        '>',
        '<',
        '>=',
        '<=',
        'between'
    ].includes(selector.op));
const isDimensionSelector = (selector)=>'object' == typeof selector && null !== selector && 'field' in selector && ('operator' in selector || 'op' in selector) && 'value' in selector && ([
        'in',
        'not in'
    ].includes(selector.operator) || [
        'in',
        'not in'
    ].includes(selector.op));
const selectByMeasure = (selector, datum)=>{
    const op = selector.operator || selector.op;
    const selectorValueArr = Array.isArray(selector.value) ? selector.value : [
        selector.value
    ];
    switch(op){
        case '=':
            if (String(datum[selector.field]) === String(selectorValueArr[0]) || nearlyEqual(Number(datum[selector.field]), Number(selectorValueArr[0]))) return true;
            break;
        case '==':
            if (datum[selector.field] === selectorValueArr[0]) return true;
            break;
        case '!=':
            if (datum[selector.field] !== selectorValueArr[0]) return true;
            break;
        case '>':
            if (datum[selector.field] > selectorValueArr[0] && !nearlyEqual(Number(datum[selector.field]), Number(selectorValueArr[0]))) return true;
            break;
        case '<':
            if (datum[selector.field] < selectorValueArr[0] && !nearlyEqual(Number(datum[selector.field]), Number(selectorValueArr[0]))) return true;
            break;
        case '>=':
            if (datum[selector.field] >= selectorValueArr[0] || nearlyEqual(Number(datum[selector.field]), Number(selectorValueArr[0]))) return true;
            break;
        case '<=':
            if (datum[selector.field] <= selectorValueArr[0] || nearlyEqual(Number(datum[selector.field]), Number(selectorValueArr[0]))) return true;
            break;
        case 'between':
            if (Array.isArray(selector.value) && (datum[selector.field] >= selectorValueArr[0] || nearlyEqual(Number(datum[selector.field]), Number(selectorValueArr[0]))) && (datum[selector.field] <= selectorValueArr[1] || nearlyEqual(Number(datum[selector.field]), Number(selectorValueArr[1])))) return true;
            break;
    }
    return false;
};
const selectByDmension = (selector, datum)=>{
    const op = selector.operator || selector.op;
    const selectorValueArr = Array.isArray(selector.value) ? selector.value : [
        selector.value
    ];
    switch(op){
        case 'in':
            if (selectorValueArr.includes(datum[selector.field])) return true;
            break;
        case 'not in':
            if (!selectorValueArr.includes(datum[selector.field])) return true;
            break;
    }
    return false;
};
const selectByPartial = (selector, datum)=>Object.keys(selector).every((key)=>datum[key] === selector[key]);
const selectByValue = (selector, datum)=>Object.values(datum).some((v)=>v === selector);
export { isDimensionSelector, isMeasureSelector, isPartialDatumSelector, isValueSelector, selectByDmension, selectByMeasure, selectByPartial, selectByValue, selector_selector as selector, selectorDatum };
