// @visactor/vseed/dist/esm/types/properties/regressionLine/zRegressionLine.js
import { z } from "zod";
const zEcdfRegressionLine = z.object({
    enable: z.boolean().nullish(),
    color: z.string().nullish(),
    lineWidth: z.number().nullish(),
    lineDash: z.array(z.number()).nullish(),
    text: z.string().nullish(),
    textColor: z.string().nullish(),
    textFontSize: z.number().nullish(),
    textFontWeight: z.number().nullish()
});
const zKdeRegressionLine = zEcdfRegressionLine.extend({});
const zLinearRegressionLine = zEcdfRegressionLine.extend({
    confidenceIntervalVisible: z.boolean().nullish(),
    confidenceIntervalColor: z.string().nullish(),
    confidenceIntervalOpacity: z.number().nullish()
});
const zLogisticRegressionLine = zLinearRegressionLine.extend({});
const zLowessRegressionLine = zLinearRegressionLine.extend({});
const zPolynomialRegressionLine = zLinearRegressionLine.extend({
    degree: z.number().nullish()
});
const zRegressionLine = z.object({
    ecdfRegressionLine: zEcdfRegressionLine.or(z.array(zEcdfRegressionLine)).nullish(),
    kdeRegressionLine: zKdeRegressionLine.or(z.array(zKdeRegressionLine)).nullish(),
    linearRegressionLine: zLinearRegressionLine.or(z.array(zLinearRegressionLine)).nullish(),
    lowessRegressionLine: zLowessRegressionLine.or(z.array(zLowessRegressionLine)).nullish(),
    polynomialRegressionLine: zPolynomialRegressionLine.or(z.array(zPolynomialRegressionLine)).nullish(),
    logisticRegressionLine: zLogisticRegressionLine.or(z.array(zLogisticRegressionLine)).nullish()
});
export { zEcdfRegressionLine, zKdeRegressionLine, zLinearRegressionLine, zLogisticRegressionLine, zLowessRegressionLine, zPolynomialRegressionLine, zRegressionLine };
