// @visactor/vseed/dist/esm/pipeline/spec/chart/pipes/annotation/utils.js
const isSubset = (sub, obj, excludeMeasuresIds)=>Object.entries(sub).every(([key, value])=>{
        if (excludeMeasuresIds && excludeMeasuresIds.includes(key)) return false;
        if ('string' == typeof value) return obj[key] === value;
        if ('number' == typeof value) return obj[key] === value;
        return true;
    });
const ANNOTATION_AREA_TEXT_STYLE_BY_POSITION = {
    top: {
        textAlign: 'center',
        textBaseline: 'top'
    },
    topRight: {
        textAlign: 'right',
        textBaseline: 'top'
    },
    topLeft: {
        textAlign: 'left',
        textBaseline: 'top'
    },
    bottom: {
        textAlign: 'center',
        textBaseline: 'bottom'
    },
    bottomLeft: {
        textAlign: 'left',
        textBaseline: 'bottom'
    },
    bottomRight: {
        textAlign: 'right',
        textBaseline: 'bottom'
    },
    left: {
        textAlign: 'left',
        textBaseline: 'middle'
    },
    right: {
        textAlign: 'right',
        textBaseline: 'middle'
    }
};
export { ANNOTATION_AREA_TEXT_STYLE_BY_POSITION, isSubset };
