// @visactor/vseed/dist/esm/pipeline/spec/chart/pipes/series/series.js
import { execPipeline } from "../../../../utils/index.js";
const series = (...args)=>{
    const result = {
        type: 'common',
        padding: 0,
        region: [
            {
                clip: true
            }
        ]
    };
    return (_, context)=>{
        result.series = args.map((pipeline)=>execPipeline(pipeline, context, {}));
        return result;
    };
};
const seriesDualAxis = (...args)=>{
    const result = {
        type: 'common',
        padding: 0,
        labelLayout: 'region',
        region: [
            {
                clip: true
            }
        ]
    };
    const createDualContext = (context, index)=>{
        const { advancedVSeed } = context;
        const dataset = advancedVSeed.dataset[index];
        return {
            ...context,
            advancedVSeed: {
                ...advancedVSeed,
                dataset: dataset
            }
        };
    };
    return (_, context)=>{
        result.series = args.map((pipeline, index)=>{
            const seriesContext = createDualContext(context, index);
            return execPipeline(pipeline, seriesContext, {});
        });
        return result;
    };
};
export { series, seriesDualAxis };
