// @visactor/vseed/dist/esm/pipeline/spec/chart/pipes/init/index.js
import { initColumn } from "./column.js";
import { initBar } from "./bar.js";
import { initBarParallel } from "./barParallel.js";
import { initArea } from "./area.js";
import { initLine } from "./line.js";
import { initColumnParallel } from "./columnParallel.js";
import { initPie } from "./pie.js";
import { initDonut } from "./donut.js";
import { initRose } from "./rose.js";
import { initRoseParallel } from "./roseParallel.js";
import { initScatter } from "./scatter.js";
import { initRadar } from "./radar.js";
import { initFunnel } from "./funnel.js";
import { initPivot } from "./pivot.js";
import { initDualAxisPrimary, initDualAxisSecondary } from "./dualAxis.js";
import { initHistogram } from "./histogram.js";
import { initBoxplot } from "./boxplot.js";
export { initArea, initBar, initBarParallel, initBoxplot, initColumn, initColumnParallel, initDonut, initDualAxisPrimary, initDualAxisSecondary, initFunnel, initHistogram, initLine, initPie, initPivot, initRadar, initRose, initRoseParallel, initScatter };
