// @visactor/vseed/dist/esm/pipeline/spec/chart/pipes/color/color.js
const color_color = (spec, context)=>{
    const result = {
        ...spec
    };
    const { advancedVSeed } = context;
    const { datasetReshapeInfo, chartType } = advancedVSeed;
    const { unfoldInfo } = datasetReshapeInfo[0];
    const baseConfig = advancedVSeed.config[chartType];
    if (!baseConfig || !baseConfig.color) return result;
    const colorItems = unfoldInfo.colorItems;
    const colorIdMap = unfoldInfo.colorIdMap;
    const { color } = baseConfig;
    const { colorScheme, colorMapping } = color;
    result.color = {
        type: 'ordinal',
        domain: colorItems,
        range: colorScheme,
        specified: createSpecifiedForColorMapping(colorMapping, colorIdMap, colorItems)
    };
    return result;
};
const createSpecifiedForColorMapping = (colorMapping, colorIdMap, colorItems)=>{
    if (!colorMapping || !colorIdMap || !colorItems) return;
    const matchedList = [];
    const colors = Object.entries(colorMapping).sort((a, b)=>b[0].length - a[0].length);
    const accurateMap = colors.reduce((prev, cur)=>{
        const name = cur[0];
        const colorValue = cur[1];
        const accurateMatchedList = Object.entries(colorIdMap).filter(([colorKey, colorObj])=>colorKey === name || colorObj.alias === name || colorObj.id === name);
        accurateMatchedList.forEach((item)=>{
            prev[item[0]] = colorValue;
            matchedList.push(name);
        });
        return prev;
    }, {});
    const fuzzyMap = colors.reduce((prev, cur)=>{
        const name = cur[0];
        const colorValue = cur[1];
        if (matchedList.includes(name)) return prev;
        const fuzzyMatchedList = Object.entries(colorIdMap).filter(([colorKey, colorObj])=>colorKey.includes(name) || colorObj.alias.includes(name) || colorObj.id.includes(name));
        fuzzyMatchedList.forEach((item)=>{
            if (prev[item[0]]) return;
            prev[item[0]] = colorValue;
        });
        return prev;
    }, {});
    return {
        ...fuzzyMap,
        ...accurateMap
    };
};
export { color_color as color, createSpecifiedForColorMapping };
