// @visactor/vseed/dist/esm/pipeline/spec/chart/pipes/color/index.js
import { color } from "./color.js";
import { colorAdapter, getColorMeasureId, isLinearColor } from "./colorAdapter.js";
import { linearColor } from "./linearColor.js";
import { linearColorForDualAxis } from "./linearColorForDualAxis.js";
import { colorBarStyleFill } from "./colorBarStyleFill.js";
import { colorLineStyleFill } from "./colorLineStyleFill.js";
import { colorPointStyleFill } from "./colorPointStyleFill.js";
import { colorAreaStyleFill } from "./colorAreaStyleFill.js";
import { colorCellStyleFill } from "./colorCellStyleFill.js";
import { colorFunnelStyleFill } from "./colorFunnelStyleFill.js";
import { colorPieStyleFill } from "./colorPieStyleFill.js";
import { colorRoseStyleFill } from "./colorRoseStyleFill.js";
import { colorBoxPlotStyleFill } from "./colorBoxPlotStyleFill.js";
export { color, colorAdapter, colorAreaStyleFill, colorBarStyleFill, colorBoxPlotStyleFill, colorCellStyleFill, colorFunnelStyleFill, colorLineStyleFill, colorPieStyleFill, colorPointStyleFill, colorRoseStyleFill, getColorMeasureId, isLinearColor, linearColor, linearColorForDualAxis };
