// @visactor/vseed/dist/esm/types/properties/config/area.js
import { z } from "zod";
import { zXBandAxis, zYLinearAxis } from "./axes/index.js";
import { zCrosshairLine } from "./crosshair/index.js";
import { zBackgroundColor } from "./backgroundColor/backgroundColor.js";
import { zColor } from "./color/color.js";
import { zLabel } from "./label/index.js";
import { zLegend } from "./legend/legend.js";
import { zTooltip } from "./tooltip/tooltip.js";
import { zAnnotationConfig } from "./annotation/zAnnotaion.js";
import { zPivotChartGridConfig } from "./pivotGrid/index.js";
import { zDimensionLinkage } from "./dimensionLinkage/dimensionLinkage.js";
const zAreaConfig = z.object({
    backgroundColor: zBackgroundColor.nullish(),
    label: zLabel.nullish(),
    color: zColor.nullish(),
    tooltip: zTooltip.nullish(),
    legend: zLegend.nullish(),
    xAxis: zXBandAxis.nullish(),
    yAxis: zYLinearAxis.nullish(),
    crosshairLine: zCrosshairLine.nullish(),
    pivotGrid: zPivotChartGridConfig.nullish(),
    annotation: zAnnotationConfig.nullish(),
    dimensionLinkage: zDimensionLinkage.nullish()
});
const zAreaPercentConfig = zAreaConfig;
export { zAreaConfig, zAreaPercentConfig };
