// @visactor/vseed/dist/esm/types/properties/config/bar.js
import { z } from "zod";
import { zXLinearAxis, zYBandAxis } from "./axes/index.js";
import { zCrosshairRect } from "./crosshair/index.js";
import { zStackCornerRadius } from "./stackCornerRadius/stackCornerRadius.js";
import { zBackgroundColor } from "./backgroundColor/backgroundColor.js";
import { zColor } from "./color/color.js";
import { zLabel } from "./label/index.js";
import { zLegend } from "./legend/legend.js";
import { zTooltip } from "./tooltip/tooltip.js";
import { zAnnotationConfig } from "./annotation/zAnnotaion.js";
import { zPivotChartGridConfig } from "./pivotGrid/index.js";
import { zDimensionLinkage } from "./dimensionLinkage/dimensionLinkage.js";
const zBarConfig = z.object({
    backgroundColor: zBackgroundColor.nullish(),
    label: zLabel.nullish(),
    color: zColor.nullish(),
    tooltip: zTooltip.nullish(),
    legend: zLegend.nullish(),
    xAxis: zXLinearAxis.nullish(),
    yAxis: zYBandAxis.nullish(),
    crosshairRect: zCrosshairRect.nullish(),
    stackCornerRadius: zStackCornerRadius.nullish(),
    pivotGrid: zPivotChartGridConfig.nullish(),
    annotation: zAnnotationConfig.nullish(),
    dimensionLinkage: zDimensionLinkage.nullish()
});
const zBarParallelConfig = zBarConfig;
const zBarPercentConfig = zBarConfig;
export { zBarConfig, zBarParallelConfig, zBarPercentConfig };
