// @visactor/vseed/dist/esm/types/properties/config/boxplot.js
import { z } from "zod";
import { zXBandAxis, zYLinearAxis } from "./axes/index.js";
import { zCrosshairRect } from "./crosshair/index.js";
import { zBackgroundColor } from "./backgroundColor/backgroundColor.js";
import { zColor } from "./color/color.js";
import { zLabel } from "./label/index.js";
import { zLegend } from "./legend/legend.js";
import { zTooltip } from "./tooltip/tooltip.js";
import { zAnnotationConfig } from "./annotation/zAnnotaion.js";
import { zPivotChartGridConfig } from "./pivotGrid/index.js";
import { zWhiskersConfig } from "./whiskers/index.js";
import { zDimensionLinkage } from "./dimensionLinkage/dimensionLinkage.js";
import { zBoxPlotStyle, zOutlierStyle } from "../markStyle/index.js";
import { zBoxGapInGroup, zBoxMaxWidth } from "./boxWidth/index.js";
const zBoxplotConfig = z.object({
    backgroundColor: zBackgroundColor.nullish(),
    label: zLabel.nullish(),
    color: zColor.nullish(),
    tooltip: zTooltip.nullish(),
    legend: zLegend.nullish(),
    xAxis: zXBandAxis.nullish(),
    yAxis: zYLinearAxis.nullish(),
    crosshairRect: zCrosshairRect.nullish(),
    pivotGrid: zPivotChartGridConfig.nullish(),
    annotation: zAnnotationConfig.nullish(),
    whiskers: zWhiskersConfig.nullish(),
    dimensionLinkage: zDimensionLinkage.nullish(),
    boxPlotStyle: zBoxPlotStyle.omit({
        selector: true
    }).nullish(),
    outlierStyle: zOutlierStyle.omit({
        selector: true
    }).nullish(),
    boxMaxWidth: zBoxMaxWidth.nullish(),
    boxGapInGroup: zBoxGapInGroup.nullish()
});
export { zBoxplotConfig };
