// @visactor/vseed/dist/esm/types/properties/config/column.js
import { z } from "zod";
import { zXBandAxis, zYLinearAxis } from "./axes/index.js";
import { zCrosshairRect } from "./crosshair/index.js";
import { zStackCornerRadius } from "./stackCornerRadius/stackCornerRadius.js";
import { zBackgroundColor } from "./backgroundColor/backgroundColor.js";
import { zColor } from "./color/color.js";
import { zLabel } from "./label/index.js";
import { zLegend } from "./legend/legend.js";
import { zTooltip } from "./tooltip/tooltip.js";
import { zAnnotationConfig } from "./annotation/zAnnotaion.js";
import { zPivotChartGridConfig } from "./pivotGrid/index.js";
import { zBarGapInGroup, zBarMaxWidth } from "./barWidth/index.js";
import { zRegressionLine } from "../regressionLine/index.js";
import { zDimensionLinkage } from "./dimensionLinkage/dimensionLinkage.js";
const zColumnParallelConfig = z.object({
    backgroundColor: zBackgroundColor.nullish(),
    label: zLabel.nullish(),
    color: zColor.nullish(),
    tooltip: zTooltip.nullish(),
    legend: zLegend.nullish(),
    xAxis: zXBandAxis.nullish(),
    yAxis: zYLinearAxis.nullish(),
    crosshairRect: zCrosshairRect.nullish(),
    stackCornerRadius: zStackCornerRadius.nullish(),
    barMaxWidth: zBarMaxWidth.nullish(),
    barGapInGroup: zBarGapInGroup.nullish(),
    pivotGrid: zPivotChartGridConfig.nullish(),
    annotation: zAnnotationConfig.nullish(),
    dimensionLinkage: zDimensionLinkage.nullish()
});
const zColumnConfig = zColumnParallelConfig.extend({
    regressionLine: zRegressionLine.nullish()
});
const zColumnPercentConfig = zColumnParallelConfig.extend({});
export { zColumnConfig, zColumnParallelConfig, zColumnPercentConfig };
