// @visactor/vseed/dist/esm/types/properties/config/config.js
import { z } from "zod";
import { zTableConfig } from "./table.js";
import { zLineConfig } from "./line.js";
import { zColumnConfig, zColumnParallelConfig, zColumnPercentConfig } from "./column.js";
import { zBarConfig, zBarParallelConfig, zBarPercentConfig } from "./bar.js";
import { zAreaConfig, zAreaPercentConfig } from "./area.js";
import { zScatterConfig } from "./scatter.js";
import { zDualAxisConfig } from "./dualAxis.js";
import { zDonutConfig, zPieConfig, zRadarConfig } from "./pie.js";
import { zRoseConfig, zRoseParallelConfig } from "./rose.js";
import { zFunnelConfig } from "./funnel.js";
import { zHeatmapConfig } from "./heatmap.js";
import { zPivotTableConfig } from "./pivotTable.js";
import { zBoxplotConfig } from "./boxplot.js";
import { zHistogramConfig } from "./histogram.js";
const zConfig = z.object({
    table: zTableConfig.nullish(),
    pivotTable: zPivotTableConfig.nullish(),
    line: zLineConfig.nullish(),
    column: zColumnConfig.nullish(),
    columnParallel: zColumnParallelConfig.nullish(),
    columnPercent: zColumnPercentConfig.nullish(),
    bar: zBarConfig.nullish(),
    barParallel: zBarParallelConfig.nullish(),
    barPercent: zBarPercentConfig.nullish(),
    area: zAreaConfig.nullish(),
    areaPercent: zAreaPercentConfig.nullish(),
    scatter: zScatterConfig.nullish(),
    dualAxis: zDualAxisConfig.nullish(),
    rose: zRoseConfig.nullish(),
    roseParallel: zRoseParallelConfig.nullish(),
    pie: zPieConfig.nullish(),
    donut: zDonutConfig.nullish(),
    radar: zRadarConfig.nullish(),
    funnel: zFunnelConfig.nullish(),
    heatmap: zHeatmapConfig.nullish(),
    boxPlot: zBoxplotConfig.nullish(),
    histogram: zHistogramConfig.nullish()
});
export { zConfig };
