// @visactor/vseed/dist/esm/types/properties/config/dualAxis.js
import { z } from "zod";
import { zXBandAxis, zYLinearAxis } from "./axes/index.js";
import { zCrosshairRect } from "./crosshair/index.js";
import { zBackgroundColor } from "./backgroundColor/backgroundColor.js";
import { zColor } from "./color/color.js";
import { zLabel } from "./label/index.js";
import { zLegend } from "./legend/legend.js";
import { zTooltip } from "./tooltip/tooltip.js";
import { zDualChartType } from "../chartType/index.js";
import { zAnnotationConfig } from "./annotation/zAnnotaion.js";
import { zPivotChartGridConfig } from "./pivotGrid/index.js";
import { zDimensionLinkage } from "./dimensionLinkage/dimensionLinkage.js";
const zDualAxisConfig = z.object({
    backgroundColor: zBackgroundColor.nullish(),
    label: zLabel.nullish(),
    color: zColor.nullish(),
    tooltip: zTooltip.nullish(),
    legend: zLegend.nullish(),
    dualChartType: z.array(zDualChartType).or(zDualChartType).nullish(),
    alignTicks: z.array(z.boolean()).or(z.boolean()).nullish(),
    primaryYAxis: z.array(zYLinearAxis).or(zYLinearAxis).nullish(),
    secondaryYAxis: z.array(zYLinearAxis).or(zYLinearAxis).nullish(),
    xAxis: zXBandAxis.nullish(),
    crosshairRect: zCrosshairRect.nullish(),
    pivotGrid: zPivotChartGridConfig.nullish(),
    annotation: zAnnotationConfig.nullish(),
    dimensionLinkage: zDimensionLinkage.nullish()
});
export { zDualAxisConfig };
