// @visactor/vseed/dist/esm/types/properties/config/histogram.js
import { z } from "zod";
import { zXBandAxis, zYLinearAxis } from "./axes/index.js";
import { zCrosshairRect } from "./crosshair/index.js";
import { zStackCornerRadius } from "./stackCornerRadius/stackCornerRadius.js";
import { zBackgroundColor } from "./backgroundColor/backgroundColor.js";
import { zColor } from "./color/color.js";
import { zLabel } from "./label/index.js";
import { zLegend } from "./legend/legend.js";
import { zTooltip } from "./tooltip/tooltip.js";
import { zAnnotationConfig } from "./annotation/zAnnotaion.js";
import { zPivotChartGridConfig } from "./pivotGrid/index.js";
import { zRegressionLine } from "../regressionLine/index.js";
import { zDimensionLinkage } from "./dimensionLinkage/dimensionLinkage.js";
const zHistogramConfig = z.object({
    backgroundColor: zBackgroundColor.nullish(),
    label: zLabel.nullish(),
    color: zColor.nullish(),
    tooltip: zTooltip.nullish(),
    legend: zLegend.nullish(),
    xAxis: zXBandAxis.nullish(),
    yAxis: zYLinearAxis.nullish(),
    crosshairRect: zCrosshairRect.nullish(),
    stackCornerRadius: zStackCornerRadius.nullish(),
    pivotGrid: zPivotChartGridConfig.nullish(),
    annotation: zAnnotationConfig.nullish(),
    binCount: z.number().positive().nullish(),
    binStep: z.number().positive().nullish(),
    binValueType: z.literal('count').or(z.literal('percentage')).nullish(),
    regressionLine: zRegressionLine.nullish(),
    dimensionLinkage: zDimensionLinkage.nullish()
});
export { zHistogramConfig };
