// @visactor/vseed/dist/esm/pipeline/spec/chart/pipes/dataset/datasetYX.js
import { isPivotChart } from "../../../../utils/index.js";
const datasetYX = (spec, context)=>{
    const { advancedVSeed, vseed } = context;
    const { analysis, datasetReshapeInfo } = advancedVSeed;
    const { unfoldInfo } = datasetReshapeInfo[0];
    const orderMapping = analysis?.orderMapping || {};
    const angle = unfoldInfo.encodingAngle;
    const y = unfoldInfo.encodingY;
    const colorId = unfoldInfo.encodingColorId;
    const id = datasetReshapeInfo[0].id;
    const fields = {};
    if (angle) fields[angle] = {
        sortIndex: 0
    };
    if (y) {
        const order = orderMapping[y];
        if (order) fields[y] = {
            sortIndex: 0,
            domain: order,
            lockStatisticsByDomain: true
        };
        else fields[y] = {
            sortIndex: 0
        };
    }
    if (colorId) {
        const order = orderMapping[colorId];
        if (order) fields[colorId] = {
            sortIndex: 0,
            domain: order,
            lockStatisticsByDomain: true
        };
        else fields[colorId] = {
            sortIndex: 0
        };
    }
    return {
        ...spec,
        data: {
            id,
            values: isPivotChart(vseed) ? void 0 : advancedVSeed.dataset,
            fields: fields
        }
    };
};
export { datasetYX };
