// @visactor/vseed/dist/esm/types/advancedVSeed.js
import { z } from "zod";
import { zChartType } from "./properties/chartType/index.js";
import { zDataset } from "./properties/dataset/index.js";
import { zDimensionTree } from "./properties/dimensions/index.js";
import { zMeasureTree } from "./properties/measures/index.js";
import { zEncoding } from "./properties/encoding/index.js";
import { zDatasetReshapeInfo } from "./properties/datasetReshapeInfo/index.js";
import { zCustomThemeConfig, zTheme } from "./properties/theme/index.js";
import { zConfig } from "./properties/config/index.js";
import { zAnalysis, zAnnotation, zMarkStyle, zRegressionLine } from "./properties/index.js";
import { zLocale } from "./i18n/index.js";
import { zCellStyle } from "./properties/cellStyle/cellStyle.js";
const zAdvancedVSeed = z.object({
    chartType: zChartType,
    dataset: zDataset,
    datasetReshapeInfo: zDatasetReshapeInfo,
    pivotAllDatasetReshapeInfo: zDatasetReshapeInfo,
    dimensions: zDimensionTree,
    measures: zMeasureTree,
    reshapeMeasures: zMeasureTree.optional(),
    reshapeDimensions: zDimensionTree.optional(),
    encoding: zEncoding,
    config: zConfig,
    analysis: zAnalysis,
    theme: zTheme,
    markStyle: zMarkStyle,
    cellStyle: zCellStyle,
    customTheme: zCustomThemeConfig,
    annotation: zAnnotation,
    locale: zLocale,
    regressionLine: zRegressionLine
});
export { zAdvancedVSeed };
