// @visactor/vseed/dist/esm/builder/register/chartType/index.js
import { registerTable } from "./table.js";
import { registerPivotTable } from "./pivotTable.js";
import { registerLine } from "./line.js";
import { registerColumn } from "./column.js";
import { registerColumnParallel } from "./columnParallel.js";
import { registerColumnPercent } from "./columnPercent.js";
import { registerBar } from "./bar.js";
import { registerBarPercent } from "./barPercent.js";
import { registerBarParallel } from "./barParallel.js";
import { registerArea } from "./area.js";
import { registerAreaPercent } from "./areaPercent.js";
import { registerScatter } from "./scatter.js";
import { registerDualAxis } from "./dualAxis.js";
import { registerPie } from "./pie.js";
import { registerDonut } from "./donut.js";
import { registerRose } from "./rose.js";
import { registerRoseParallel } from "./roseParallel.js";
import { registerRadar } from "./radar.js";
import { registerHeatmap } from "./heatmap.js";
import { registerFunnel } from "./funnel.js";
import { registerBoxPlot } from "./boxplot.js";
import { registerHistogram } from "./histogram.js";
export { registerArea, registerAreaPercent, registerBar, registerBarParallel, registerBarPercent, registerBoxPlot, registerColumn, registerColumnParallel, registerColumnPercent, registerDonut, registerDualAxis, registerFunnel, registerHeatmap, registerHistogram, registerLine, registerPie, registerPivotTable, registerRadar, registerRose, registerRoseParallel, registerScatter, registerTable };
