// @visactor/vseed/dist/esm/pipeline/advanced/chart/pipes/config/index.js
import { lineConfig } from "./line.js";
import { columnConfig } from "./column.js";
import { pieConfig } from "./pie.js";
import { dualAxisConfig } from "./dualAxis.js";
import { scatterConfig } from "./scatter.js";
import { histogramConfig } from "./histogram.js";
import { boxplotConfig } from "./boxplot.js";
const areaConfig = lineConfig;
const areaPercentConfig = lineConfig;
const barConfig = columnConfig;
const barParallelConfig = columnConfig;
const barPercentConfig = columnConfig;
const columnParallelConfig = columnConfig;
const columnPercentConfig = columnConfig;
const donutConfig = pieConfig;
const roseConfig = pieConfig;
const roseParallelConfig = pieConfig;
const radarConfig = pieConfig;
const funnelConfig = pieConfig;
const heatmapConfig = pieConfig;
export { areaConfig, areaPercentConfig, barConfig, barParallelConfig, barPercentConfig, boxplotConfig, columnConfig, columnParallelConfig, columnPercentConfig, donutConfig, dualAxisConfig, funnelConfig, heatmapConfig, histogramConfig, lineConfig, pieConfig, radarConfig, roseConfig, roseParallelConfig, scatterConfig };
