// @visactor/vseed/dist/esm/pipeline/utils/chatType.js
import { measureDepth } from "./measures/index.js";
import { isMeasureTreeWithChildren, isMeasureTreeWithParentId } from "../advanced/chart/pipes/measures/utils.js";
import { unique } from "remeda";
import { ChartTypeEnum, DEFAULT_PARENT_ID } from "./constant.js";
const isTable = (vseed)=>vseed.chartType === ChartTypeEnum.Table;
const isPivotTable = (vseed)=>vseed.chartType === ChartTypeEnum.PivotTable;
const isRadar = (vseed)=>vseed.chartType === ChartTypeEnum.Radar;
const isAreaPercent = (vseed)=>vseed.chartType === ChartTypeEnum.AreaPercent;
const isColumnPercent = (vseed)=>vseed.chartType === ChartTypeEnum.ColumnPercent;
const isBarPercent = (vseed)=>vseed.chartType === ChartTypeEnum.BarPercent;
const isBarLikeChart = (vseed)=>vseed.chartType === ChartTypeEnum.Bar || vseed.chartType === ChartTypeEnum.BarPercent || vseed.chartType === ChartTypeEnum.BarParallel;
const isVTable = (vseed)=>[
        'table',
        'pivotTable'
    ].includes(vseed.chartType);
const isVChart = (vseed)=>!isVTable(vseed);
const isPivotChart = (vseed)=>{
    if (isVTable(vseed)) return false;
    if (isPivot(vseed)) return true;
    return isCombination(vseed);
};
const isPivot = (vseed)=>{
    const { dimensions = [] } = vseed;
    return dimensions && dimensions.some((dimension)=>'row' === dimension.encoding || 'column' === dimension.encoding);
};
const isCombination = (vseed)=>{
    if (isMeasureTreeWithParentId(vseed.measures)) {
        const parentIds = vseed.measures?.map((measure)=>measure.parentId || DEFAULT_PARENT_ID);
        return parentIds && unique(parentIds).length > 1;
    }
    if ('dualAxis' === vseed.chartType || 'scatter' === vseed.chartType) {
        if ('scatter' === vseed.chartType) {
            if (isMeasureTreeWithChildren(vseed.measures)) {
                const depth = measureDepth(vseed.measures);
                return 3 === depth;
            }
            if (vseed.scatterMeasures && vseed.scatterMeasures.length > 1) return true;
            return false;
        }
        if ('dualAxis' === vseed.chartType) {
            if (isMeasureTreeWithChildren(vseed.measures)) {
                const depth = measureDepth(vseed.measures);
                return 3 === depth;
            }
            if (vseed.dualMeasures && vseed.dualMeasures.length > 1) return true;
        }
        return false;
    }
    const { measures = [] } = vseed;
    const hasMeasureGroup = measures && measures.some((measure)=>measure && measure.children);
    return hasMeasureGroup;
};
export { isAreaPercent, isBarLikeChart, isBarPercent, isColumnPercent, isCombination, isPivot, isPivotChart, isPivotTable, isRadar, isTable, isVChart, isVTable };
