// @visactor/vseed/dist/esm/pipeline/utils/constant.js
const ANNOTATION_Z_INDEX = 1000;
const LINEAR_AXIS_INNER_OFFSET_TOP = 12;
const DUAL_AXIS_CHART_COLUMN_Z_INDEX = 1000;
const DUAL_AXIS_CHART_NON_COLUMN_Z_INDEX = 1001;
const DUAL_AXIS_LABEL_Z_INDEX = 1002;
const DEFAULT_PARENT_ID = '__DefaultParentId__';
const AXIS_LABEL_SPACE = 8;
const BAND_AXIS_INNER_OFFSET_IN_PIVOT = 2;
const ChartTypeEnum = {
    Table: 'table',
    PivotTable: 'pivotTable',
    Line: 'line',
    Column: 'column',
    ColumnPercent: 'columnPercent',
    ColumnParallel: 'columnParallel',
    Bar: 'bar',
    BarPercent: 'barPercent',
    BarParallel: 'barParallel',
    Area: 'area',
    AreaPercent: 'areaPercent',
    DualAxis: 'dualAxis',
    Scatter: 'scatter',
    Rose: 'rose',
    RoseParallel: 'roseParallel',
    Pie: 'pie',
    Donut: 'donut',
    Radar: 'radar',
    Funnel: 'funnel',
    Heatmap: 'heatmap',
    Boxplot: 'boxPlot',
    Histogram: 'histogram'
};
const DATUM_HIDE_KEY = '__VSEED_HideItem__';
export { ANNOTATION_Z_INDEX, AXIS_LABEL_SPACE, BAND_AXIS_INNER_OFFSET_IN_PIVOT, ChartTypeEnum, DATUM_HIDE_KEY, DEFAULT_PARENT_ID, DUAL_AXIS_CHART_COLUMN_Z_INDEX, DUAL_AXIS_CHART_NON_COLUMN_Z_INDEX, DUAL_AXIS_LABEL_Z_INDEX, LINEAR_AXIS_INNER_OFFSET_TOP };
