// @visactor/vseed/dist/esm/types/properties/config/legend/legend.js
import { z } from "zod";
const zLegend = z.object({
    enable: z.boolean().default(true).nullish(),
    border: z.boolean().default(true).nullish(),
    maxSize: z.number().default(1).nullish(),
    shapeType: z["enum"]([
        'circle',
        'cross',
        'diamond',
        'square',
        'arrow',
        'arrow2Left',
        'arrow2Right',
        'wedge',
        'thinTriangle',
        'triangle',
        'triangleUp',
        'triangleDown',
        'triangleRight',
        'triangleLeft',
        'stroke',
        'star',
        'wye',
        'rect',
        'arrowLeft',
        'arrowRight',
        'rectRound',
        'roundLine'
    ]).default('rectRound').nullish(),
    position: z["enum"]([
        'left',
        'leftTop',
        'leftBottom',
        'lt',
        'lb',
        'top',
        'topLeft',
        'topRight',
        'tl',
        'tr',
        'right',
        'rightTop',
        'rightBottom',
        'rt',
        'rb',
        'bottom',
        'bottomLeft',
        'bottomRight',
        'bl',
        'br'
    ]).default('bottom').nullish(),
    labelColor: z.string().default('#fff').nullish(),
    pagerIconColor: z.string().nullish(),
    pagerIconDisableColor: z.string().nullish(),
    labelFontSize: z.number().default(12).nullish(),
    labelFontWeight: z.number().or(z.string()).default(400).nullish()
});
const zColorLegend = z.object({
    position: z["enum"]([
        'left',
        'leftTop',
        'leftBottom',
        'lt',
        'lb',
        'top',
        'topLeft',
        'topRight',
        'tl',
        'tr',
        'right',
        'rightTop',
        'rightBottom',
        'rt',
        'rb',
        'bottom',
        'bottomLeft',
        'bottomRight',
        'bl',
        'br'
    ]).default('bottom').nullish(),
    enable: z.boolean().default(true).nullish()
});
export { zColorLegend, zLegend };
