// @visactor/vseed/dist/esm/pipeline/spec/chart/pipes/crosshair/verticalCrosshairRect.js
const verticalCrosshairRect = (spec, context)=>{
    const result = {
        ...spec
    };
    const { advancedVSeed, vseed } = context;
    const { chartType } = vseed;
    const config = advancedVSeed.config?.[chartType]?.crosshairRect;
    if (!config) return result;
    if (!result.crosshair) result.crosshair = {};
    const visible = config.visible || true;
    const rectColor = config.rectColor || void 0;
    const labelColor = config.labelColor || void 0;
    const labelVisible = config.labelVisible || false;
    const labelBackgroundColor = config.labelBackgroundColor || void 0;
    const crosshair = result.crosshair;
    crosshair.xField = {
        visible,
        line: {
            type: 'rect',
            style: {
                lineWidth: 0,
                opacity: 0.26,
                fill: rectColor
            }
        },
        label: {
            visible: labelVisible,
            labelBackground: {
                visible: labelVisible,
                style: {
                    fill: labelBackgroundColor
                }
            },
            style: {
                fill: labelColor
            }
        }
    };
    const xAxisConfig = result.axes?.find((v)=>'bottom' === v.orient);
    const xAxisFormatter = xAxisConfig?.label?.formatMethod;
    if (xAxisFormatter) crosshair.xField.label.formatMethod = (text)=>xAxisFormatter(text);
    return result;
};
export { verticalCrosshairRect };
