// @visactor/vseed/dist/esm/pipeline/spec/table/pipes/theme/bodyStyle.js
const bodyStyle = (spec, context)=>{
    const result = {
        ...spec
    };
    const { advancedVSeed } = context;
    const { chartType, config } = advancedVSeed;
    const themeConfig = config?.[chartType];
    if (!result.theme || !themeConfig) return result;
    const borderColor = themeConfig.borderColor || 'rgb(224, 224, 224)';
    const backgroundColor = themeConfig.bodyBackgroundColor || '#fff';
    const fontColor = themeConfig.bodyFontColor || '#1B1F23';
    const fontSize = themeConfig.bodyFontSize || 12;
    const hoverCellBgColor = themeConfig.hoverBodyBackgroundColor || '#bedaff';
    const hoverInlineColor = themeConfig.hoverBodyInlineBackgroundColor || '#bedaff';
    if (!result.theme) result.theme = {};
    result.theme.bodyStyle = {
        borderColor: borderColor,
        borderLineWidth: 1,
        padding: [
            8,
            12,
            8,
            12
        ],
        textAlign: 'right',
        hover: {
            cellBgColor: hoverCellBgColor,
            inlineRowBgColor: hoverInlineColor,
            inlineColumnBgColor: hoverInlineColor
        },
        color: fontColor,
        fontSize: fontSize,
        fontStyle: 'normal',
        fontWeight: 'normal',
        fontVariant: 'normal',
        bgColor: backgroundColor,
        lineHeight: 1.5 * fontSize
    };
    return result;
};
export { bodyStyle };
