// @visactor/vseed/dist/esm/pipeline/spec/table/pipes/theme/headerStyle.js
const headerStyle = (spec, context)=>{
    const result = {
        ...spec
    };
    const { advancedVSeed } = context;
    const { config, chartType } = advancedVSeed;
    const themConfig = config?.[chartType];
    if (!result.theme || !themConfig) return result;
    const borderColor = themConfig.borderColor || 'rgb(224, 224, 224)';
    const backgroundColor = themConfig.headerBackgroundColor || '#EEF1F5';
    const fontColor = themConfig.headerFontColor || '#1B1F23';
    const fontSize = themConfig.headerFontSize || 12;
    const hoverCellBgColor = themConfig.hoverHeaderBackgroundColor || '#bedaff';
    const hoverInlineColor = themConfig.hoverHeaderInlineBackgroundColor || '#bedaff';
    result.theme.headerStyle = {
        borderColor: borderColor,
        borderLineWidth: 1,
        padding: [
            8,
            12,
            8,
            12
        ],
        textAlign: 'center',
        hover: {
            cellBgColor: hoverCellBgColor,
            inlineRowBgColor: hoverInlineColor,
            inlineColumnBgColor: hoverInlineColor
        },
        frameStyle: {
            borderColor: borderColor,
            borderLineWidth: [
                0,
                0,
                1,
                0
            ]
        },
        fontSize: fontSize,
        fontVariant: 'normal',
        fontStyle: 'normal',
        fontWeight: 'bold',
        color: fontColor,
        bgColor: backgroundColor,
        lineHeight: 1.5 * fontSize
    };
    return result;
};
export { headerStyle };
